tools/workload-simulator.html

```diff
@@ -0,0 +1,1160 @@
+        gpuPresent = webGPUCanvas.getContext('gpupresent');
+        swapChain = gpuPresent.configureSwapChain({device, format: swapChainFormat});
+        const bindGroupLayout = device.createBindGroupLayout({
+          entries: [
+            {
+              binding: 0,
+              // TODO: These shouldn't need to be visible to both stages. There's a bug in Dawn that
+              // incorrectly requires variables that are unused in one stage to be visible to all stages.
+              visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
+              buffer: { type: 'uniform', },
+            },
+            {
+              binding: 1,
+              visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
+              sampler: { type: 'filtering', },
+            },
+            {
+              binding: 2,
+              visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
+              texture: { sampleType: 'float', },
+            },
+          ],
+        });
+        const pipelineLayout = device.createPipelineLayout({
+          bindGroupLayouts: [bindGroupLayout],
+        });
+        if (multisampling.checked) {
+          multisampleRenderAttachment = device.createTexture({
+            size: { width: size, height: size },
+            sampleCount,
+            format: swapChainFormat,
+            usage: GPUTextureUsage.RENDER_ATTACHMENT,
+          })
+          if (!multisampleRenderAttachment)
+            errorMessage.textContent = 'Failed to allocate multisample render attachment.';
+        }
+        let shaderModule = device.createShaderModule({ code: `
+          let pos : array<vec2<f32>, 6> = array<vec2<f32>, 6>(
+              vec2<f32>(-1., 1.),
+              vec2<f32>(-1., -1.),
+              vec2<f32>(1., -1.),
+              vec2<f32>(-1., 1.),
+              vec2<f32>(1., -1.),
+              vec2<f32>(1., 1.));
+
+          [[block]] struct Uniforms {
+            xy: vec2<f32>;
+          };
+          [[binding(0), group(0)]] var<uniform> uniforms: Uniforms;
+          [[binding(1), group(0)]] var sampler1: sampler;
+          [[binding(2), group(0)]] var texture1: texture_2d<f32>;
+
+          struct Output {
+            [[builtin(position)]] Position : vec4<f32>;
+            [[location(0)]] uv : vec2<f32>;
+          };
+          
+          [[stage(vertex)]]
+          fn vs([[builtin(vertex_index)]] VertexIndex : i32) -> Output {
+            var output : Output;
+            output.Position = vec4<f32>(pos[VertexIndex] / vec2<f32>(2.,2.) + uniforms.xy, 0.0, 1.0);
+            output.uv = pos[VertexIndex] / vec2<f32>(2., -2.) + vec2<f32>(.5, .5);
+            return output;
+          }
+
+          [[stage(fragment)]]
+          fn fs([[location(0)]] uv: vec2<f32>) -> [[location(0)]] vec4<f32> {
+            return textureSample(texture1, sampler1, uv);
+          }`, });
+        if (shaderModule.compilationInfo)
```

Comment:
no need put it all together, adding at least an empty line here would help readability

Reply:
Agreed.